### DISCUSSION TITLE
Toggles for `reportIncompatibleVariableOverride` and covariant mutable overrides

### DISCUSSION THREAD

**Discussion by hauntsaninja** [microsoft/pyright#6872] [Q&A]
Context: https://github.com/microsoft/pyright/issues/5933

I'm curious if there's a way to only ignore warnings for covariant mutable overrides, or if disabling all of `reportIncompatibleVariableOverride` is the only option (assuming I'm not willing to disallow covariant mutable overrides)

**Comment by erictraut** [comment]
No, pyright doesn't differentiate between covariant mutable overrides and other non-type-safe overrides. You can use a `# type: ignore` or `# pyright: ignore` on a case-by-case basis.

Out of curiosity, why do you want to single out just that one type of behavior?

**Reply by hauntsaninja** [reply]
Thanks!

Several projects at work use pyright. Those projects tried to upgrade pyright and started getting many, many new warnings because they picked up #5955 (edit: might be a follow up change causing most of new warnings), so I'm figuring out what their options are.

In general, [I find this diagnostic to be disruptive and low signal](https://github.com/python/mypy/issues/3208#issuecomment-1719106888). I've yet to see it catch a bug in real world code, so if it were possible I would disable it so as to remove a source of nuisance (both ongoing and during the one-time upgrade).

Admittedly, I'm more accustomed to mypy which goes pretty far in the direction of giving users fine-grained control over diagnostics.

**Reply by hmc-cs-mdrissi** [reply]
My experience is similar. My work codebase is smaller at ~100k lines that manually fixing them was doable, but mutable overrides while unsafe I didn't notice a real bug caused by any of them.

In practice many of overrides were for variables intended as immutable in parent, but defining something as immutable is a bit messy with how properties vs attributes vs cached properties mix today. This [thread](https://discuss.python.org/t/need-a-way-to-type-hint-attributes-that-is-compatible-with-duck-typing/41137) currently lacks a simple answer. I'd like a way to define,

```python
class Foo(Protocol):
  readable x: float
```

where

```python
from functools import cached_property

class Foo1:
  def __init__(self):
    self.x: int = 3

class Foo2:
  @property
  def x(self) -> int:
    ...

class Foo3:
  @cached_property
  def x(self) -> int:
    ...
```

all satisfy Foo protocol/fine to inherit from it.

are all compatible. Using `@property` as a way to make mutable vs immutable is rather awkward when override that uses attribute/cached_property appears. Sadly cached_property is not even a subtype of property although I've never had a real bug caused by this and just type: ignore.

**Reply by elenakrittik** [reply]
This is also often happens with [`textual`](https://github.com/Textualize/textual) due to overriding `App.SCREENS` which is a `dict`, but is practically never modified. `textual` authors' intent is also very clear given that `SCREENS` are not modified internally too (there are only read in `App.__init__` to `App._installed_screens` once, then screen management is done via helper methods that only modify `App._installed_screens`). Currently, aside from `# type: ignore` of course, the only way is to fully duplicate `SCREENS`'s signature in the subclass. Ability to silence exactly this error somewhere in `pyproject.toml` will allow to document why the error is insignificant while not ignoring other possible issues with code.

**Comment by vors** [comment]
I did a binary search and the change we saw on one of the files from our codebase is between versions `338` and `339`. @erictraut I looked at the changes and the most suspicious one was https://github.com/microsoft/pyright/commit/703215a42a7f34c1944fb38bea10f72d2e5a5ff8 -- does it change only the TypedDict or other inheritances too?

**Reply by hauntsaninja** [reply]
Specifically on the following file with `pydantic==1.10.13`:
```
import pydantic
from typing import Literal

class BaseState(pydantic.BaseModel):
    name: str

class Idle(BaseState):
    name: Literal["Idle"] = "Idle"
```
pyright 1.1.339 complains about:
```
/Users/shantanu/devoai/openai/xx.py
  /Users/shantanu/devoai/openai/xx.py:12:5 - error: "name" overrides symbol of same name in class "BaseState"
    Variable is mutable so its type is invariant
      Override type "Literal['Idle']" is not the same as base type "str" (reportIncompatibleVariableOverride)
1 error, 0 warnings, 0 informations 
```
pyright is correct to complain about this, assuming you care about covariant mutable overrides. The trouble is I don't think we do :D
